test/qd8-f16-qb4w-gemm-minmax.cc

```diff
@@ -0,0 +1,494 @@
+#include <string>
+#include <vector>
+
+#include <xnnpack/allocator.h>
+#include <xnnpack/common.h>
+#include <xnnpack/gemm.h>
+#include <xnnpack/igemm.h>
+#include <xnnpack/isa-checks.h>
+#include <xnnpack/microparams-init.h>
+#include <xnnpack/pack.h>
+#include <xnnpack/ppmm.h>
+#include <xnnpack/requantization.h>
+
+#include "gemm-microkernel-tester.h"
+#include <gtest/gtest.h>
+
+namespace {
+
+std::vector<GemmTestParams> CreateTests1(
+    size_t k_block, size_t adj_k_block,
+    size_t mr, size_t nr, size_t kr, size_t sr,
+    bool is_igemm,
+    std::function<void(GemmMicrokernelTester& tester)> test_func,
+    std::function<void()> isa_check = nullptr) {
+  std::string kbs = std::to_string(k_block);
+  std::string kb2s = std::to_string(k_block * 2);
+  std::string akbs = std::to_string(adj_k_block);
+  std::string nrs = std::to_string(nr);
+
+  std::vector<GemmTestParams> gemm_tests;
+  gemm_tests.reserve(42);
+
+  gemm_tests.push_back(GemmTestParams(
+      "k_eq_" + kbs,
+      GemmMicrokernelTester()
+          .mr(mr).nr(nr).kr(kr).sr(sr).m(mr).n(nr).k(k_block)
+          .b_zero_point(8)
+          .bl(kr * sr * 2)
+      , test_func, isa_check));
+  gemm_tests.push_back(GemmTestParams(
+      "strided_cn",
+      GemmMicrokernelTester()
+          .mr(mr).nr(nr).kr(kr).sr(sr).m(mr).n(nr).k(k_block)
+          .cn_stride(NextPrime(nr + 1))
+          .b_zero_point(8)
+          .bl(kr * sr * 2)
+    , test_func, isa_check));
+  if (!is_igemm) {
+    gemm_tests.push_back(GemmTestParams(
+        "k_eq_" + kbs + "_strided_a",
+        GemmMicrokernelTester()
+            .mr(mr).nr(nr).kr(kr).sr(sr).m(mr).n(nr).k(k_block)
+            .a_stride(NextPrime(k_block + 1))
+            .b_zero_point(8)
+            .bl(kr * sr * 2)
+        , test_func, isa_check));
+  }
+  gemm_tests.push_back(GemmTestParams(
+      "k_eq_" + kbs + "_subtile",
+      GemmMicrokernelTester()
+          .mr(mr).nr(nr).kr(kr).sr(sr).k(k_block).iterations(1)
+          .b_zero_point(8)
+          .bl(kr * sr * 2)
+      , test_func, isa_check)
+      .loop_n(1, nr)
+      .loop_m(1, mr));
+  gemm_tests.push_back(GemmTestParams(
+      "k_eq_" + kbs + "_subtile_m",
+      GemmMicrokernelTester()
+          .mr(mr).nr(nr).kr(kr).sr(sr).n(nr).k(k_block).iterations(1)
+          .b_zero_point(8)
+          .bl(kr * sr * 2)
+      , test_func, isa_check)
+      .loop_m(1, mr));
+  gemm_tests.push_back(GemmTestParams(
+      "k_eq_" + kbs + "_subtile_n",
+      GemmMicrokernelTester()
+          .mr(mr).nr(nr).kr(kr).sr(sr).m(mr).k(k_block).iterations(1)
+          .b_zero_point(8)
+          .bl(kr * sr * 2)
+      , test_func, isa_check)
+      .loop_n(1, nr));
+  if (k_block > 1) {
+    gemm_tests.push_back(GemmTestParams(
+        "k_lt_" + akbs,
+        GemmMicrokernelTester()
+            .mr(mr).nr(nr).kr(kr).sr(sr).m(mr).n(nr)
+            .b_zero_point(8)
+            .bl(kr * sr * 2)
+        , test_func, isa_check)
+        .loop_k(1, adj_k_block - 1));
```

Comment:
so this will not work with `kr > 1` or `sr > 1`. Because you will loop in K dim in steps of 1 which will break the constraints. Same for all the +1 steps. `loop_k` where you step in multiple of `block_k` or `adj_k_block` i.e. full blocks should be ok. Not commenting on all the places.

Easiest is to hack a kr=2 kernel for scalar and see what breaks.

Or alternatively we can upstream this and then trim it later when we introduce fancier kernels.

Reply:
Good catch. This is a little bit of a pain to handle cleanly in test gen. Are we committed to not handling partial-k in kernels? If so, I'll figure something out.

Reply:
Discussed offline. I will update test gen.

Reply:
Do you know how many we skip in this case?